### ISSUE TITLE
[1.3.x turbo] deadlock in sbt internal classloaders when running tests

### ISSUE THREAD

**Issue by bjaglin** [sbt/sbt#5116]
## steps
* sbt version
  - reproducable on 1.3.0 & 1.3.2
  - I do not remember seeing it on early 1.3 RCs (before https://github.com/sbt/sbt/commit/9c821b7b13eb14d20eedf51efa28a8f82bdb22e6#diff-9426c31e9b78333e024e94f62b057583 was introduced) but I can't tell for sure
  - never reproduced on 1.2.x
* turbo mode enabled
* default test settings: no fork, parallel execution

## problem

`sbt test` hangs sometimes  (~10% of the builds on our 16 vCPU CI hosts)

## expectation

`sbt test` fails or succeeds, but does not hang

## notes

I haven't been able to extract a test suite to exhibit the problem, and this big multi-module project is not open-source... Here are the relevant bits of a thread dump when sbt hangs:

```
2019-09-23 09:34:28 Found one Java-level deadlock:
2019-09-23 09:34:28 =============================
2019-09-23 09:34:28 "pool-22-thread-2-ScalaTest-running-TeadsCommitmentSpec":
2019-09-23 09:34:28   waiting to lock monitor 0x00007f46a43c7600 (object 0x00000007e7d80018, a sbt.internal.ReverseLookupClassLoaderHolder$BottomClassLoader),
2019-09-23 09:34:28   which is held by "pool-22-thread-12-ScalaTest-running-ContractSpec"
2019-09-23 09:34:28 "pool-22-thread-12-ScalaTest-running-ContractSpec":
2019-09-23 09:34:28   waiting to lock monitor 0x00007f46a43c8200 (object 0x00000007e7b00090, a sbt.internal.ReverseLookupClassLoaderHolder$ReverseLookupClassLoader),
2019-09-23 09:34:28   which is held by "pool-22-thread-6-ScalaTest-running-ContractPartiesSpec"
2019-09-23 09:34:28 "pool-22-thread-6-ScalaTest-running-ContractPartiesSpec":
2019-09-23 09:34:28   waiting to lock monitor 0x00007f47d800b800 (object 0x00000007e7b002f0, a java.lang.Object),
2019-09-23 09:34:28   which is held by "pool-22-thread-12-ScalaTest-running-ContractSpec"


2019-09-23 09:34:28 "pool-22-thread-2-ScalaTest-running-TeadsCommitmentSpec" #2110 prio=5 os_prio=0 cpu=1079.84ms elapsed=1377.98s tid=0x00007f48c582d000 nid=0xe77 waiting for monitor entry  [0x00007f46f3ffc000]
2019-09-23 09:34:28    java.lang.Thread.State: BLOCKED (on object monitor)
2019-09-23 09:34:28 	at tv.teads.domains.mrg_contract.commands.models.ContractOpportunityQuarterlyRollover$.create(ContractOpportunityQuarterlyRollover.scala:18)
2019-09-23 09:34:28 	at tv.teads.domains.mrg_contract.commands.models.TeadsCommitmentSpec.$anonfun$new$3(TeadsCommitmentSpec.scala:14)
...

09-23 09:34:28 "pool-22-thread-12-ScalaTest-running-ContractSpec" #2120 prio=5 os_prio=0 cpu=1600.77ms elapsed=1377.85s tid=0x00007f48c6ad3800 nid=0xe81 waiting for monitor entry  [0x00007f47327f6000]
2019-09-23 09:34:28    java.lang.Thread.State: BLOCKED (on object monitor)
2019-09-23 09:34:28 	at java.lang.ClassLoader.loadClass(java.base@11.0.2/ClassLoader.java:568)
2019-09-23 09:34:28 	- waiting to lock <0x00000007e7b00090> (a sbt.internal.ReverseLookupClassLoaderHolder$ReverseLookupClassLoader)
2019-09-23 09:34:28 	at sbt.internal.ReverseLookupClassLoaderHolder$ReverseLookupClassLoader.super$loadClass(LayeredClassLoaders.scala:171)
2019-09-23 09:34:28 	at sbt.internal.ReverseLookupClassLoaderHolder$ReverseLookupClassLoader.$anonfun$loadClass$1(LayeredClassLoaders.scala:171)
2019-09-23 09:34:28 	at sbt.internal.ReverseLookupClassLoaderHolder$ReverseLookupClassLoader$$Lambda$9569/0x000000080258e040.apply(Unknown Source)
2019-09-23 09:34:28 	at sbt.internal.ReverseLookupClassLoaderHolder$ClassLoadingLock.withLock(LayeredClassLoaders.scala:116)
2019-09-23 09:34:28 	- locked <0x00000007e7b002f0> (a java.lang.Object)
2019-09-23 09:34:28 	at sbt.internal.ReverseLookupClassLoaderHolder$ReverseLookupClassLoader.loadClass(LayeredClassLoaders.scala:171)
2019-09-23 09:34:28 	at sbt.internal.ReverseLookupClassLoaderHolder$BottomClassLoader.liftedTree1$1(LayeredClassLoaders.scala:236)
2019-09-23 09:34:28 	at sbt.internal.ReverseLookupClassLoaderHolder$BottomClassLoader.loadClass(LayeredClassLoaders.scala:236)
2019-09-23 09:34:28 	at java.lang.ClassLoader.loadClass(java.base@11.0.2/ClassLoader.java:521)
2019-09-23 09:34:28 	at tv.teads.domains.mrg_contract.commands.models.ContractSpec.$anonfun$new$4(ContractSpec.scala:33)
...

019-09-23 09:34:28 "pool-22-thread-6-ScalaTest-running-ContractPartiesSpec" #2114 prio=5 os_prio=0 cpu=1502.60ms elapsed=1377.86s tid=0x00007f48c57b6800 nid=0xe7b waiting for monitor entry  [0x00007f4753fea000]
2019-09-23 09:34:28    java.lang.Thread.State: BLOCKED (on object monitor)
2019-09-23 09:34:28 	at sbt.internal.ReverseLookupClassLoaderHolder$ClassLoadingLock.withLock(LayeredClassLoaders.scala:116)
2019-09-23 09:34:28 	- waiting to lock <0x00000007e7b002f0> (a java.lang.Object)
2019-09-23 09:34:28 	at sbt.internal.ReverseLookupClassLoaderHolder$ReverseLookupClassLoader.loadClass(LayeredClassLoaders.scala:171)
2019-09-23 09:34:28 	at sbt.internal.ReverseLookupClassLoaderHolder$ReverseLookupClassLoader.loadClass(LayeredClassLoaders.scala:185)
2019-09-23 09:34:28 	at java.lang.ClassLoader.loadClass(java.base@11.0.2/ClassLoader.java:521)
2019-09-23 09:34:28 	at java.lang.ClassLoader.defineClass1(java.base@11.0.2/Native Method)
2019-09-23 09:34:28 	at java.lang.ClassLoader.defineClass(java.base@11.0.2/ClassLoader.java:1016)
2019-09-23 09:34:28 	at java.security.SecureClassLoader.defineClass(java.base@11.0.2/SecureClassLoader.java:174)
2019-09-23 09:34:28 	at java.net.URLClassLoader.defineClass(java.base@11.0.2/URLClassLoader.java:550)
2019-09-23 09:34:28 	at java.net.URLClassLoader$1.run(java.base@11.0.2/URLClassLoader.java:458)
2019-09-23 09:34:28 	at java.net.URLClassLoader$1.run(java.base@11.0.2/URLClassLoader.java:452)
2019-09-23 09:34:28 	at java.security.AccessController.doPrivileged(java.base@11.0.2/Native Method)
2019-09-23 09:34:28 	at java.net.URLClassLoader.findClass(java.base@11.0.2/URLClassLoader.java:451)
2019-09-23 09:34:28 	at sbt.internal.ManagedClassLoader.findClass(LayeredClassLoaders.scala:376)
2019-09-23 09:34:28 	at java.lang.ClassLoader.loadClass(java.base@11.0.2/ClassLoader.java:588)
2019-09-23 09:34:28 	- locked <0x00000007e7b00090> (a sbt.internal.ReverseLookupClassLoaderHolder$ReverseLookupClassLoader)
2019-09-23 09:34:28 	at sbt.internal.ReverseLookupClassLoaderHolder$ReverseLookupClassLoader.super$loadClass(LayeredClassLoaders.scala:171)
2019-09-23 09:34:28 	at sbt.internal.ReverseLookupClassLoaderHolder$ReverseLookupClassLoader.$anonfun$loadClass$1(LayeredClassLoaders.scala:171)
2019-09-23 09:34:28 	at sbt.internal.ReverseLookupClassLoaderHolder$ReverseLookupClassLoader$$Lambda$9569/0x000000080258e040.apply(Unknown Source)
2019-09-23 09:34:28 	at sbt.internal.ReverseLookupClassLoaderHolder$ClassLoadingLock.withLock(LayeredClassLoaders.scala:116)
2019-09-23 09:34:28 	- locked <0x00000007e7b19090> (a java.lang.Object)
2019-09-23 09:34:28 	at sbt.internal.ReverseLookupClassLoaderHolder$ReverseLookupClassLoader.loadClass(LayeredClassLoaders.scala:171)
2019-09-23 09:34:28 	at sbt.internal.ReverseLookupClassLoaderHolder$ReverseLookupClassLoader.loadClass(LayeredClassLoaders.scala:185)
2019-09-23 09:34:28 	at java.lang.ClassLoader.loadClass(java.base@11.0.2/ClassLoader.java:521)
2019-09-23 09:34:28 	at java.lang.ClassLoader.defineClass1(java.base@11.0.2/Native Method)
2019-09-23 09:34:28 	at java.lang.ClassLoader.defineClass(java.base@11.0.2/ClassLoader.java:1016)
2019-09-23 09:34:28 	at java.security.SecureClassLoader.defineClass(java.base@11.0.2/SecureClassLoader.java:174)
2019-09-23 09:34:28 	at java.net.URLClassLoader.defineClass(java.base@11.0.2/URLClassLoader.java:550)
2019-09-23 09:34:28 	at java.net.URLClassLoader$1.run(java.base@11.0.2/URLClassLoader.java:458)
2019-09-23 09:34:28 	at java.net.URLClassLoader$1.run(java.base@11.0.2/URLClassLoader.java:452)
2019-09-23 09:34:28 	at java.security.AccessController.doPrivileged(java.base@11.0.2/Native Method)
2019-09-23 09:34:28 	at java.net.URLClassLoader.findClass(java.base@11.0.2/URLClassLoader.java:451)
2019-09-23 09:34:28 	at sbt.internal.ManagedClassLoader.findClass(LayeredClassLoaders.scala:376)
2019-09-23 09:34:28 	at java.lang.ClassLoader.loadClass(java.base@11.0.2/ClassLoader.java:588)
2019-09-23 09:34:28 	- locked <0x00000007e7b00090> (a sbt.internal.ReverseLookupClassLoaderHolder$ReverseLookupClassLoader)
2019-09-23 09:34:28 	at sbt.internal.ReverseLookupClassLoaderHolder$ReverseLookupClassLoader.super$loadClass(LayeredClassLoaders.scala:171)
2019-09-23 09:34:28 	at sbt.internal.ReverseLookupClassLoaderHolder$ReverseLookupClassLoader.$anonfun$loadClass$1(LayeredClassLoaders.scala:171)
2019-09-23 09:34:28 	at sbt.internal.ReverseLookupClassLoaderHolder$ReverseLookupClassLoader$$Lambda$9569/0x000000080258e040.apply(Unknown Source)
2019-09-23 09:34:28 	at sbt.internal.ReverseLookupClassLoaderHolder$ClassLoadingLock.withLock(LayeredClassLoaders.scala:116)
2019-09-23 09:34:28 	- locked <0x00000007e7b1c568> (a java.lang.Object)
2019-09-23 09:34:28 	at sbt.internal.ReverseLookupClassLoaderHolder$ReverseLookupClassLoader.loadClass(LayeredClassLoaders.scala:171)
2019-09-23 09:34:28 	at sbt.internal.ReverseLookupClassLoaderHolder$ReverseLookupClassLoader.loadClass(LayeredClassLoaders.scala:185)
2019-09-23 09:34:28 	at java.lang.ClassLoader.loadClass(java.base@11.0.2/ClassLoader.java:521)
2019-09-23 09:34:28 	at java.lang.ClassLoader.defineClass1(java.base@11.0.2/Native Method)
2019-09-23 09:34:28 	at java.lang.ClassLoader.defineClass(java.base@11.0.2/ClassLoader.java:1016)
2019-09-23 09:34:28 	at java.security.SecureClassLoader.defineClass(java.base@11.0.2/SecureClassLoader.java:174)
2019-09-23 09:34:28 	at java.net.URLClassLoader.defineClass(java.base@11.0.2/URLClassLoader.java:550)
2019-09-23 09:34:28 	at java.net.URLClassLoader$1.run(java.base@11.0.2/URLClassLoader.java:458)
2019-09-23 09:34:28 	at java.net.URLClassLoader$1.run(java.base@11.0.2/URLClassLoader.java:452)
2019-09-23 09:34:28 	at java.security.AccessController.doPrivileged(java.base@11.0.2/Native Method)
2019-09-23 09:34:28 	at java.net.URLClassLoader.findClass(java.base@11.0.2/URLClassLoader.java:451)
2019-09-23 09:34:28 	at sbt.internal.ManagedClassLoader.findClass(LayeredClassLoaders.scala:376)
2019-09-23 09:34:28 	at java.lang.ClassLoader.loadClass(java.base@11.0.2/ClassLoader.java:588)
2019-09-23 09:34:28 	- locked <0x00000007e7b00090> (a sbt.internal.ReverseLookupClassLoaderHolder$ReverseLookupClassLoader)
2019-09-23 09:34:28 	at sbt.internal.ReverseLookupClassLoaderHolder$ReverseLookupClassLoader.super$loadClass(LayeredClassLoaders.scala:171)
2019-09-23 09:34:28 	at sbt.internal.ReverseLookupClassLoaderHolder$ReverseLookupClassLoader.$anonfun$loadClass$1(LayeredClassLoaders.scala:171)
2019-09-23 09:34:28 	at sbt.internal.ReverseLookupClassLoaderHolder$ReverseLookupClassLoader$$Lambda$9569/0x000000080258e040.apply(Unknown Source)
2019-09-23 09:34:28 	at sbt.internal.ReverseLookupClassLoaderHolder$ClassLoadingLock.withLock(LayeredClassLoaders.scala:116)
2019-09-23 09:34:28 	- locked <0x00000007e7b20620> (a java.lang.Object)
2019-09-23 09:34:28 	at sbt.internal.ReverseLookupClassLoaderHolder$ReverseLookupClassLoader.loadClass(LayeredClassLoaders.scala:171)
2019-09-23 09:34:28 	at sbt.internal.ReverseLookupClassLoaderHolder$ReverseLookupClassLoader.loadClass(LayeredClassLoaders.scala:185)
2019-09-23 09:34:28 	at java.lang.ClassLoader.loadClass(java.base@11.0.2/ClassLoader.java:521)
2019-09-23 09:34:28 	at java.lang.ClassLoader.defineClass1(java.base@11.0.2/Native Method)
2019-09-23 09:34:28 	at java.lang.ClassLoader.defineClass(java.base@11.0.2/ClassLoader.java:1016)
2019-09-23 09:34:28 	at java.security.SecureClassLoader.defineClass(java.base@11.0.2/SecureClassLoader.java:174)
2019-09-23 09:34:28 	at java.net.URLClassLoader.defineClass(java.base@11.0.2/URLClassLoader.java:550)
2019-09-23 09:34:28 	at java.net.URLClassLoader$1.run(java.base@11.0.2/URLClassLoader.java:458)
2019-09-23 09:34:28 	at java.net.URLClassLoader$1.run(java.base@11.0.2/URLClassLoader.java:452)
2019-09-23 09:34:28 	at java.security.AccessController.doPrivileged(java.base@11.0.2/Native Method)
2019-09-23 09:34:28 	at java.net.URLClassLoader.findClass(java.base@11.0.2/URLClassLoader.java:451)
2019-09-23 09:34:28 	at sbt.internal.ManagedClassLoader.findClass(LayeredClassLoaders.scala:376)
2019-09-23 09:34:28 	at java.lang.ClassLoader.loadClass(java.base@11.0.2/ClassLoader.java:588)
2019-09-23 09:34:28 	- locked <0x00000007e7b00090> (a sbt.internal.ReverseLookupClassLoaderHolder$ReverseLookupClassLoader)
2019-09-23 09:34:28 	at sbt.internal.ReverseLookupClassLoaderHolder$ReverseLookupClassLoader.super$loadClass(LayeredClassLoaders.scala:171)
2019-09-23 09:34:28 	at sbt.internal.ReverseLookupClassLoaderHolder$ReverseLookupClassLoader.$anonfun$loadClass$1(LayeredClassLoaders.scala:171)
2019-09-23 09:34:28 	at sbt.internal.ReverseLookupClassLoaderHolder$ReverseLookupClassLoader$$Lambda$9569/0x000000080258e040.apply(Unknown Source)
2019-09-23 09:34:28 	at sbt.internal.ReverseLookupClassLoaderHolder$ClassLoadingLock.withLock(LayeredClassLoaders.scala:116)
2019-09-23 09:34:28 	- locked <0x00000007e7b24708> (a java.lang.Object)
2019-09-23 09:34:28 	at sbt.internal.ReverseLookupClassLoaderHolder$ReverseLookupClassLoader.loadClass(LayeredClassLoaders.scala:171)
2019-09-23 09:34:28 	at sbt.internal.ReverseLookupClassLoaderHolder$ReverseLookupClassLoader.loadClass(LayeredClassLoaders.scala:185)
2019-09-23 09:34:28 	at java.lang.ClassLoader.loadClass(java.base@11.0.2/ClassLoader.java:521)
2019-09-23 09:34:28 	at java.lang.ClassLoader.defineClass1(java.base@11.0.2/Native Method)
2019-09-23 09:34:28 	at java.lang.ClassLoader.defineClass(java.base@11.0.2/ClassLoader.java:1016)
2019-09-23 09:34:28 	at java.security.SecureClassLoader.defineClass(java.base@11.0.2/SecureClassLoader.java:174)
2019-09-23 09:34:28 	at java.net.URLClassLoader.defineClass(java.base@11.0.2/URLClassLoader.java:550)
2019-09-23 09:34:28 	at java.net.URLClassLoader$1.run(java.base@11.0.2/URLClassLoader.java:458)
2019-09-23 09:34:28 	at java.net.URLClassLoader$1.run(java.base@11.0.2/URLClassLoader.java:452)
2019-09-23 09:34:28 	at java.security.AccessController.doPrivileged(java.base@11.0.2/Native Method)
2019-09-23 09:34:28 	at java.net.URLClassLoader.findClass(java.base@11.0.2/URLClassLoader.java:451)
2019-09-23 09:34:28 	at sbt.internal.ManagedClassLoader.findClass(LayeredClassLoaders.scala:376)
2019-09-23 09:34:28 	at java.lang.ClassLoader.loadClass(java.base@11.0.2/ClassLoader.java:588)
2019-09-23 09:34:28 	- locked <0x00000007e7b00090> (a sbt.internal.ReverseLookupClassLoaderHolder$ReverseLookupClassLoader)
2019-09-23 09:34:28 	at sbt.internal.ReverseLookupClassLoaderHolder$ReverseLookupClassLoader.super$loadClass(LayeredClassLoaders.scala:171)
2019-09-23 09:34:28 	at sbt.internal.ReverseLookupClassLoaderHolder$ReverseLookupClassLoader.$anonfun$loadClass$1(LayeredClassLoaders.scala:171)
2019-09-23 09:34:28 	at sbt.internal.ReverseLookupClassLoaderHolder$ReverseLookupClassLoader$$Lambda$9569/0x000000080258e040.apply(Unknown Source)
2019-09-23 09:34:28 	at sbt.internal.ReverseLookupClassLoaderHolder$ClassLoadingLock.withLock(LayeredClassLoaders.scala:116)
2019-09-23 09:34:28 	- locked <0x00000007e7b250f0> (a java.lang.Object)
2019-09-23 09:34:28 	at sbt.internal.ReverseLookupClassLoaderHolder$ReverseLookupClassLoader.loadClass(LayeredClassLoaders.scala:171)
2019-09-23 09:34:28 	at sbt.internal.ReverseLookupClassLoaderHolder$ReverseLookupClassLoader.loadClass(LayeredClassLoaders.scala:185)
2019-09-23 09:34:28 	at java.lang.ClassLoader.loadClass(java.base@11.0.2/ClassLoader.java:521)
2019-09-23 09:34:28 	at java.lang.ClassLoader.defineClass1(java.base@11.0.2/Native Method)
2019-09-23 09:34:28 	at java.lang.ClassLoader.defineClass(java.base@11.0.2/ClassLoader.java:1016)
2019-09-23 09:34:28 	at java.security.SecureClassLoader.defineClass(java.base@11.0.2/SecureClassLoader.java:174)
2019-09-23 09:34:28 	at java.net.URLClassLoader.defineClass(java.base@11.0.2/URLClassLoader.java:550)
2019-09-23 09:34:28 	at java.net.URLClassLoader$1.run(java.base@11.0.2/URLClassLoader.java:458)
2019-09-23 09:34:28 	at java.net.URLClassLoader$1.run(java.base@11.0.2/URLClassLoader.java:452)
2019-09-23 09:34:28 	at java.security.AccessController.doPrivileged(java.base@11.0.2/Native Method)
2019-09-23 09:34:28 	at java.net.URLClassLoader.findClass(java.base@11.0.2/URLClassLoader.java:451)
2019-09-23 09:34:28 	at sbt.internal.ManagedClassLoader.findClass(LayeredClassLoaders.scala:376)
2019-09-23 09:34:28 	at java.lang.ClassLoader.loadClass(java.base@11.0.2/ClassLoader.java:588)
2019-09-23 09:34:28 	- locked <0x00000007e7b00090> (a sbt.internal.ReverseLookupClassLoaderHolder$ReverseLookupClassLoader)
2019-09-23 09:34:28 	at sbt.internal.ReverseLookupClassLoaderHolder$ReverseLookupClassLoader.super$loadClass(LayeredClassLoaders.scala:171)
2019-09-23 09:34:28 	at sbt.internal.ReverseLookupClassLoaderHolder$ReverseLookupClassLoader.$anonfun$loadClass$1(LayeredClassLoaders.scala:171)
2019-09-23 09:34:28 	at sbt.internal.ReverseLookupClassLoaderHolder$ReverseLookupClassLoader$$Lambda$9569/0x000000080258e040.apply(Unknown Source)
2019-09-23 09:34:28 	at sbt.internal.ReverseLookupClassLoaderHolder$ClassLoadingLock.withLock(LayeredClassLoaders.scala:116)
2019-09-23 09:34:28 	- locked <0x00000007e7b383c8> (a java.lang.Object)
2019-09-23 09:34:28 	at sbt.internal.ReverseLookupClassLoaderHolder$ReverseLookupClassLoader.loadClass(LayeredClassLoaders.scala:171)
2019-09-23 09:34:28 	at sbt.internal.ReverseLookupClassLoaderHolder$ReverseLookupClassLoader.loadClass(LayeredClassLoaders.scala:185)
2019-09-23 09:34:28 	at java.lang.ClassLoader.loadClass(java.base@11.0.2/ClassLoader.java:521)
2019-09-23 09:34:28 	at cats.data.NonEmptyListInstances.<init>(NonEmptyList.scala:514)
2019-09-23 09:34:28 	at cats.data.NonEmptyList$.<init>(NonEmptyList.scala:439)
2019-09-23 09:34:28 	at cats.data.NonEmptyList$.<clinit>(NonEmptyList.scala)
2019-09-23 09:34:28 	at tv.teads.domains.mrg_contract.commands.models.ContractParties$.create(ContractParties.scala:20)
...
```

**Comment by bjaglin** [comment]
I am trying to reproduce without the `turbo` flag, and will update if I manage.

**Comment by eed3si9n** [comment]
@bjaglin Thanks for the report. Is there any project that would reproduce this behavior available publicly?

**Comment by bjaglin** [comment]
@eed3si9n unfortunately, no - still trying to narrow the problem down so that I could provide a test suite, but the transient nature of the problem makes it hard.

**Comment by eatkins** [comment]
Thanks for linking to the commit that you suspect might have introduced the issue. Looking at it, it does seem like we might be holding the lock too long. That being said, turbo mode enables `ClassLoaderLayeringStrategy.AllLibraryJar` which isn't optimal if java reflection is used by dependencies to load classes from your project (which is often caused by java serialization/deserialization). I suspect that is what is happening in your project. There are two issues that I see related to this issue:
1. The deadlock itself (which is hopefully fixable)
2. The lack of feedback that java reflection was used. There is absolutely no benefit to turbo mode if java reflection is used in the dependency layer. In that scenario, it probably makes sense to hardcode the layering strategy to `ClassLoaderLayeringStrategy.ScalaLibrary`.

Point 2 can be addressed by adding a warning message that gets printed if java reflection is used in the dependency layer. We still need a simple repro for the deadlock, but I have an idea for how to generate one if my hunch about the underlying cause is correct.

**Comment by leonardehrenfried** [comment]
I'm seeing something very similar in our integration tests but it's very hard to pin down since it only happens on our CI hosts (Gitlab CI on kubernetes runners).

In our case it only happens when loading resources (application.conf via typesafe config).

About 10% of the time the test run just deadlocks and never finishes. In another 20% it cannot find the application.conf file on the classpath and the test fails.

I've also tried to narrow it down to a minimal example but so far I've failed. Whenever I thought that I have an example the problem disappears. My guess is that it's a race condition.

I've tried with and without turbo mode and it doesn't make a difference.

**Comment by bjaglin** [comment]
@leonardehrenfried you are probably experiencing another issue, as I have confirmed my problem was related to the classpath layering exposed behind the turbo flag.

**Comment by bjaglin** [comment]
@eatkins Regarding serialization: I can't tell for sure as the project has many dependencies, but the classes showing in the stack trace are generated from https://github.com/scalapb/ScalaPB, which as far as I know does not make use of reflection. We are using structural types in one place, but it does not seems related. I'll try to get rid of it and reproduce the problem. Do you have anything in mind that I could do/run to see if it's related to serialization?

**Comment by bjaglin** [comment]
One thing I could have mentioned in the description since it might impact the classloading behavior is that this is multi-project build (a simple tree with a few commons module), and the deadlock happens when running ITs on the tree leaves.

**Comment by leonardehrenfried** [comment]
I hate to keep interjecting but I have a possible clue about when this problem appears (at least in my case):

It appears to be triggered when loading a type from the same package which is in another module.

In case it isn't clear, I'm referring to the following file structure:

```
module-a/src/main/scala/com/company/A.scala
module-b/src/main/scala/com/company/B.scala
module-b/src/it/scala/com/company/BSpec.scala
```
BSpec loads B which load A. 

~If I move the files into another package my deadlocks go away.~ I spoke too soon: it makes the deadlock occur less frequently.

@bjaglin Does your project have a similar structure?

**Comment by eatkins** [comment]
@leonardehrenfried I appreciate your chiming in. I'm less convinced now that is due to java serialization than it was before. I have a simple repro that can cause deadlock without java serialization.

**Comment by leonardehrenfried** [comment]
I also just noticed that it does not solve the problem after all. It just makes it appear less frequently.

**Comment by eatkins** [comment]
I think the issue may not have been specific to turbo mode. It turns out that the sbt classloaders were not [parallel capable](https://docs.oracle.com/javase/8/docs/api/java/lang/ClassLoader.html#registerAsParallelCapable--). In https://github.com/sbt/sbt/pull/5131, I reimplemented the classloaders in java and registered them all as parallel capable. Prior to this change, I had a good repro that would deadlock almost every time. After the java switch, I don't see anymore deadlocks.

**Comment by bjaglin** [comment]
@leonardehrenfried 

> I hate to keep interjecting but I have a possible clue about when this problem appears (at least in my case):

Apologies if I sounded sharp, your troubleshooting contributions are definitely welcome!

> @bjaglin Does your project have a similar structure?

I do have classes from the same package spread across modules.

**Comment by bjaglin** [comment]
Regarding `turbo` being (or not) a factor to my issue: since I disabled `turbo`, I have not witnessed a single deadlock (out of ~100 CI builds), which obviously doesn't mean I never will, but the reproducibility rate without it is (at least) an order of magnitude lower than with it. I guess turbo might not be a cause but rather a catalyst as it increases concurrency...

@eatkins I'll see if I can build your PR locally to confirm if it helps in my case (or use a nightly if it gets merged soon).

**Comment by bjaglin** [comment]
Just an update: for some reason (CI slave load reducing the level of concurrency?), I am having a hard time reproducing the problem in the same condition as originally reported, so I cannot yet confirm whether #5131 fixed my issue or not.

**Comment by bjaglin** [comment]
Since the tentative fix was [backported to 1.3.3](https://github.com/sbt/sbt/pull/5149), we have just bumped and re-enabled `turbo` on our project (which has discussed earlier might not be necessary, but in our case at the very least increased the chances of seeing the problem), so I should be able to tell if it helps within a few days.

**Comment by bjaglin** [comment]
Not reproducible with 1.3.3, thanks!
